### ISSUE TITLE
About Collator

### ISSUE THREAD

**Issue by kimwongyuda** [TIGER-AI-Lab/VLM2Vec#35]
I really appreciate your work and am curious of collator.

In the function "_get_batch_inputs" of the collator class, items are processed through the processor one by one like below code. Each item is processed by for loop and they are padded and aggregated through torch._C._nn.pad_sequence. Then attention mask is generated by input_ids.ne.

```
for example in examples:
            text, image = example[text_idx], example[image_idx]

            if image is None:
                image_mask.append(0)
                if self.model_args.model_backbone == "llava_next":
                    inputs = self.processor(images=None, text=text, return_tensors="pt")
                elif self.model_args.model_backbone == "qwen":
                    inputs = self.processor(text=[text], images=None, return_tensors="pt",
                                            max_length=self.data_args.max_len, truncation=True)
                else:
                    inputs = self.processor(text=text, images=None, return_tensors="pt",
                                            max_length=self.data_args.max_len, truncation=True)
                input_ids.append(inputs["input_ids"].squeeze(0).unsqueeze(1))
            else:
                image_mask.append(1)
                image_exist = True
                if self.model_args.model_backbone == "llava_next":
                    inputs = self.processor(images=image, text=text, return_tensors="pt")
                    pixel_values.append(inputs['pixel_values'])
                elif self.model_args.model_backbone == "qwen":
                    inputs = self.processor(text=[text], images=[image], return_tensors="pt",
                                            max_length=self.data_args.max_len, truncation=False)
                    pixel_values.append(inputs['pixel_values'].unsqueeze(0))
                else:
                    inputs = self.processor(text=text, images=[image], return_tensors="pt",
                                            max_length=self.data_args.max_len, truncation=True)
                    pixel_values.append(inputs['pixel_values'])
                input_ids.append(inputs["input_ids"].squeeze(0).unsqueeze(1))
                if "image_sizes" in inputs:
                    image_sizes.append(inputs['image_sizes'])
                if "image_grid_thw" in inputs:
                    image_grid_thw.append(inputs['image_grid_thw'])

        input_ids = torch._C._nn.pad_sequence(
            input_ids, batch_first=True, padding_value=self.processor.tokenizer.pad_token_id
        ).squeeze(2)
        attention_mask = input_ids.ne(self.processor.tokenizer.pad_token_id)
```

However, I think items can be processed all at once if using below code. This code maybe leads to same results compared to original code with saving time.

```
def _get_batch_inputs(self, examples, text_idx, image_idx):
        txts = []
        imgs = []
        image_mask = []

        for example in examples:
            txt = example[text_idx]
            img = example[image_idx]

            if txt is not None:
                txts.append(txt)
            if img is not None:
                imgs.append(img)
                image_mask.append(1)
            else:
                image_mask.append(0)

        if self.model_args.model_backbone == "phi3_v" or self.model_args.model_backbone == "phi35v" or self.model_args.model_backbone == "llava_next":

            inputs = self.processor(images=imgs if len(imgs) > 0 else None, text=txts, return_tensors="pt", padding=True, max_length=800, truncation=True)
            inputs = {
                'input_ids': inputs['input_ids'],
                'attention_mask': inputs['attention_mask'],
                'pixel_values': inputs['pixel_values'] if 'pixel_values' in inputs else None,
                'image_sizes': inputs['image_sizes'] if 'image_sizes' in inputs else None,
            }

        elif self.model_args.model_backbone == "qwen":
            
            inputs = self.processor(images=imgs if len(imgs) > 0 else None, text=txts, return_tensors="pt", padding=True, max_length=600, truncation=True)
            inputs = {
                'input_ids':inputs['input_ids'],
                'attention_mask': inputs['attention_mask'],
                'pixel_values': inputs['pixel_values'] if 'pixel_values' in inputs else None,
                'image_grid_thw': inputs['image_grid_thw'] if 'image_grid_thw' in inputs else None,
            }

        if image_mask:
            inputs['image_mask'] = torch.Tensor(image_mask)
```

Do you have any Intent about the original codes which process items one by one?

Also, with the original codes, only right padding can be used and we can't use left padding even though llava_next and qwen model originally use left padding. What do you think of that? I am just wondering !

Thank you.

**Comment by memray** [comment]
Hi @kimwongyuda , thanks for the suggestion! Since data points with and without images are used in the same batch, we ran into a lot of troubles. Thus we opt for processing them one by one. As we know most models do not handle batch inputs very well. You may notice our latest code (late-process) has refactored the pipeline by calling process_fn [seperately](https://github.com/TIGER-AI-Lab/VLM2Vec/blob/db2af21ea3468b1b0138ee2782d3e21df93595c9/src/trainer.py#L636). Please check it out!
